### ISSUE TITLE
Proxy Transparency

### ISSUE THREAD

**Issue by seanmonstar** [runconduit/conduit#41]
# goal

A goal of the proxy is to be transparent, meaning that the application need not change behavior to benefit from the proxy. Transparency is primarily achieved by using iptables(8) to redirect traffic into the proxy.

- all tcp traffic enters the proxy on a single port.
- support for h2, http, and raw tcp transport

### HTTP1

An incoming HTTP/1 request has the format of `method request-target HTTP/1.{0,1}\r\n`. Unfortunately, that means that with a really large request target, it can require multiple reads and buffering before finally knowing if this is an HTTP/1 request.

It’s likely that non-HTTP1 data will be quick to identify, since as soon as it doesn’t match the shape of 3 words, we can assume its not HTTP.

A question that arises is whether slightly invalid HTTP requests should be classified as raw TCP instead. If the target route only speaks one of those options, the answer is simple, but if it speaks both, the answer is unresolved.

### HTTP2

We have this working as the required default at the moment. With transparency, we could know incoming is HTTP2 by peeking the socket for the `PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n` magic bytes header (or from ALPN in the TLS detection).

### TCP

If an accepted connection does not parse as either HTTP1 or HTTP2, then we may assume it’s plain TCP. There is an unresolved question around whether to accept the plain TCP or to reject as an invalid HTTP request, see the HTTP1 section.

### TLS

We can peek the socket looking for the start of a TLS record to determine if TLS should be used. This peek should happen before the HTTP/TCP detection happens.

On a brand new connection, we can assume that the first TLS record is a handshake, and thus can peek for these bytes: 

- `0x16 0x03 0xNN 0xNN 0xNN 0x01` - We’re looking for a `HANDSHAKE` byte (`0x16`), TLS version (3.N), some length we don’t yet care about, and then the `CLIENT_HELLO` byte (`0x01`).
- We could decide to be more lenient or more strict with the version check and the length bytes.

Another question here is whether we want to assume that a slightly broken TLS record should just be considered plain TCP, or should we send an error back about the bad TLS format.

## connect-side

Knowing what protocols to route to a destination likely requires some prior knowledge, retrieved by the Destination service from the controller.

Eventually, if no hint can be provided by the controller (a user hasn’t configured the service with such a hint), we can try some defaults: use HTTP1 with an Upgrade header to allow upgrading to HTTP2.

Would this default mean that without a prior knowledge hint from the controller, we assume the destination does not want plain TCP or TLS?

### Prior knowledge from controller

The destination service would be updated to report back hints about what protocols a destination supports. A proposed new destination service:

```protobuf
service Destination {
  // Given a destination, return all addresses in that destination as a long-
  // running stream of updates.
  rpc Get(common.Destination) returns (stream Update) {}
}

message Update {
  oneof update {
    Add add = 1;
    Remove remove = 2;
  }
}

message Remove {
  repeated common.TcpAddress addrs = 1;
  repeated Protocol protocols = 2;
}

message Add {
  repeated WeightedAddr addrs = 1;
  repeated Protocol protocols = 2;
}

message WeightedAddr {
  common.TcpAddress addr = 1;
  uint32 weight = 3; 
}

enum Protocol {
  HTTP_1 = 0; // should we differentiate between 1.0 and 1.1?
  HTTP_2 = 1;
  TCP = 3;
  TLS = 4; // should we include TLS versions?
}
```

These hints can be stored in Kubernetes [metadata annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) in the Endpoints API.

# Work to be done

Here is a list of tasks to be done to acheive this goal:

## Accept
  - [ ] Peek new connections for TLS records
  - [ ] Peek new connections for HTTP2 preface
  - [ ] Peek new connections for HTTP1 start-line
    - [ ] Include an HTTP1 server (hyper?)
  - [ ] Forward connections that don't match any protocol as plain TCP

## Connect
  - [ ] Update controller Destination service to include Protocol hints
    - [ ] Document how to configure protocols
  - [ ] Use protocol hints in proxy
    - [ ] Add an HTTP/1 client (hyper?)
    - [ ] Create a TCP forwarding client
    - [ ] Add TLS to proxy Connect
  - [ ] Default to HTTP/1 with Upgrade to H2 if no prior knowledge
    - [ ] Get HTTP/1 with upgrades client (hyper?)

**Comment by briansmith** [comment]
> use HTTP1 with an Upgrade header to allow upgrading to HTTP2.

Is any application actually using HTTP1 with an Upgrade header to HTTP2? I am skeptical that it is worthwhile to spend time on this unless we see some evidence it is actually done in the real world. I've literally never seen it done IRL.

**Comment by briansmith** [comment]
> accept-side

Let's distinguish between the two cases of accept-side traffic: public traffic from outside the pod, and private traffic from the service within the same pod that the proxy is in. I agree that the plan makes sense w.r.t. the private traffic we accept from the service within the same pod as the proxy. And I agree in the short term it is OK for the public traffic from outside the pod. However, in the long term I think the plan makes less sense.

In particular, the long-term plan is to provide automatic TLS and to enforce that TLS is used by default. That means that, in the long term, absent any annotations, we need to reject any non-TLS public traffic. While the annotation to opt into accepting non-secure public traffic could be a boolean flag, if the user has to add an annotation for plaintext traffic to work anyway, we might as well make the annotation value be the name of the plaintext protocol: HTTP or whatever. Then we wouldn't have to do the sniffing on the public listener, which would be better for security, predictability, and reliability.

**Comment by briansmith** [comment]
@seanmonstar What should we do with this issue? GitHub says 0/13 tasks have been done but that seems far from accurate: new tasks were identified and many tasks were already done. Should we just close this in favor of having a separate GitHub issue for each currently-outstanding issue?

**Comment by seanmonstar** [comment]
I'd say this could be closed, as it is not an accurate record of how this was implemented.
